### PR TITLE
Self-contained JAR with OS-Arch specific native libraries

### PR DESCRIPTION
**Pull request by vkuzmin-uber**
### Summary:
Improve current fat jar to make it completely self-contained JAR with OS-Arch specific native libraries.
New "fat" JAR has embedded resources with JNI library, Neuropod Core and Backend packages at com/uber/neuropod/native/OS-ARCH
Issue #457 

### Test Plan:
Tensorflow and Torchscript Java tests depends on this self-contained JAR.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,6 @@
 *.profraw
 *.pyc
 source/compile_commands.json
-source/python/.idea
 source/python/neuropod.egg-info/
```
Comment (mincomp) [inline .gitignore:4]: why remove this?

File: source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java [hunk 2]
```diff
@@ -66,32 +96,101 @@ private static boolean loadSharedLibrary() {
...
         } catch (IOException e) {
-            System.out.println(e.getMessage());
+            LOGGER.log(Level.WARNING, e.getMessage());
+            result = false;
+        }
+        return result;
+    }
+
+    private static boolean installFramework(String libAbsPath, String frameworkPackage, String neuropodBaseDir) {
+        return startShellCommand(String.format("tar -xf %s/%s -C %s", libAbsPath, frameworkPackage, neuropodBaseDir));
```
Comment (mincomp) [inline source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java:171]: this is going to take some time + CPU usage at runtime. is this okay?

File: source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java [hunk 3]
```diff
@@ -66,32 +96,101 @@ private static boolean loadSharedLibrary() {
...
+        LOGGER.log(Level.WARNING, "Command {0}", builder.command());
+        File logErr = new File("/tmp/log_err");
+        builder.redirectError(logErr);
+        File log = new File("/tmp/log");
+        builder.redirectOutput(log);
+
+        try {
+            Process process = builder.start();
+            int exitCode = process.waitFor();
+            LOGGER.log(Level.WARNING, "Exit code {0}", exitCode);
```
Comment (sepo-eng) [inline source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java:188]: Should we warn for non failure exit codes?

**Review by sepo-eng — COMMENTED** [review]
LGTM, thanks for making this change for us OSX users 👍 .

Leaving to Vivek for final approval

File: source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java (hunk 3 above)
Comment (vkuzmin-uber) [reply source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java:188]: Yes, I found it today when did self-review. I think it was late already when I created PR and I skipped several obvious issues. Anyway, thanks!

Comment (vkuzmin-uber) [reply source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java:188]: I decided to log command and exit code as Info, that's it.

File: source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java [hunk 4]
```diff
@@ -66,32 +96,101 @@ private static boolean loadSharedLibrary() {
...
+            String neuropodBaseDir = System.getenv(NEUROPOD_BASE_DIR_ENV_VAR);
+            if (neuropodBaseDir == null) {
+                neuropodBaseDir = NEUROPOD_BASE_DIR_DEFAULT;
+            }
+            for (String libName : EMBEDDED_BACKEND_PACKAGES_NAMES) {
+                File embeddedPackageFile = extractFile(libAbsPath, resPath, libName);
+                if (embeddedPackageFile != null) {
+                    LOGGER.log(Level.INFO, "Extracted embedded package file {0}", embeddedPackageFile);
+                    String fileName = embeddedPackageFile.getName();
+                    if (fileName.equals(TENSORFLOW_BACKEND) || fileName.equals(TORCHSCRIPT_BACKEND)) {
```
[the remark this answers is not in the document]
Comment (vkuzmin-uber) [reply source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java:150]: I added it in initial code where packages were in EMBEDDED_LIB_NAMES. Then I decided to extract it to separate list and didn't remove it. Found it today as well when did self-review. Anyway, thanks. Fixed!

File: source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java (hunk 2 above)
Comment (vkuzmin-uber) [reply source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java:171]: Yes, I measured and extract+load  may take up to 10s on my mac. Good that this happens once only when we initialize API (so at start of service)

File: .gitignore (hunk 1 above)
Comment (vkuzmin-uber) [reply .gitignore:4]: There is another line below .idea that match all appearances of .idea in repo.

**Comment by vkuzmin-uber** [comment]
I see Bazel Buildifier Lint failed. Found that there is a tool with linter and formatter of bazel BUILD files. https://github.com/bazelbuild/buildtools/tree/master/buildifier . Fixed.

**Comment by VivekPanyam** [comment]
I took a look, but I want to spend some time thinking about whether we can come up with a better approach if that's okay.

The point of the recent backend loading refactor was to avoid having to ship backends in a package for each language Neuropod was used from (e.g. pip packages, fat jars, etc).

The approach in this PR has some drawbacks including being limited to one version of TF and one version of Torch.

I'm going to comment some alternatives and more general thoughts on the issue you created for this PR

**Review by VivekPanyam — CHANGES_REQUESTED** [review]
See https://github.com/uber/neuropod/issues/457#issuecomment-732023918 for more details

TL;DR: I think bundling backends into the final JAR and untarring at runtime is specific to your usecase and not very useful outside the constraints of the environment you're running in. It seems like we can solve that problem in a better way by doing it in an internal "installer" library outside of Neuropod.

File: source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java (hunk 2 above)
Comment (vkuzmin-uber) [reply source/neuropod/bindings/java/src/main/java/com/uber/neuropod/LibraryLoader.java:167]: Decided to remove "backend" part from this PR for now and pack "core" only.

**Comment by vkuzmin-uber** [comment]
> See [#457 (comment)](https://github.com/uber/neuropod/issues/457#issuecomment-732023918) for more details
> 
> TL;DR: I think bundling backends into the final JAR and untarring at runtime is specific to your usecase and not very useful outside the constraints of the environment you're running in. It seems like we can solve that problem in a better way by doing it in an internal "installer" library outside of Neuropod.

I removed "pack backend tar" into Fat Jar part and left only Core native bins in specific directories (for OSX and Linux). Please review.

File: source/neuropod/bindings/java/BUILD [hunk 5]
```diff
@@ -45,6 +45,17 @@ filegroup(
...
+    name = "tar_backends",
+    srcs = [
+        "//neuropod/backends/tensorflow:neuropod_tensorflow_backend",
+        "//neuropod/backends/torchscript:neuropod_torchscript_backend",
+    ],
+    visibility = [
+        "//neuropod:__subpackages__",
+    ],
+)
+
```
Comment (VivekPanyam) [inline source/neuropod/bindings/java/BUILD:58]: No longer necessary

File: source/neuropod/bindings/java/README.md [hunk 6]
```diff
@@ -1,6 +1,30 @@
...
+rm -rf "/tmp/.neuropod_test_base" && mkdir "/tmp/.neuropod_test_base"
+bazel test --action_env=NEUROPOD_TENSORFLOW_VERSION=1.15.0 --action_env=NEUROPOD_TORCH_VERSION=1.4.0 --test_lang_filters="java" --test_tag_filters="-gpu" //...
+```
+
+Note that test builds fat jar first, that has "neuropod standard" directory structure
+```text
+0.2.0/backends/tensorflow_1.15.0/
+0.2.0/backends/torchscript_1.4.0/
+```
+that is expected by "register backends" logic.
```
Comment (VivekPanyam) [inline source/neuropod/bindings/java/README.md:27]: This is no longer accurate, please update

File: source/neuropod/bindings/java/BUILD [hunk 7]
```diff
@@ -95,8 +148,8 @@ java_library(
     data = [
         ":backends",
```
Comment (VivekPanyam) [inline source/neuropod/bindings/java/BUILD:149]: This is no longer necessary (I was going to remove it in another PR, but I think it makes sense to do here). Can you also remove the definition of `backends` above please?

File: source/neuropod/bindings/java/BUILD [hunk 8]
```diff
@@ -75,15 +97,46 @@ filegroup(
...
+mkdir -p $$jni_path
+mv libneuropod_jni.so libneuropod.so neuropod_multiprocess_worker $$jni_path
+cd ..
+jar cf $@ -C tmp/ .  # Creates the new jar file.
+""",
+    }),
+)
+
+java_binary(
+    name = "neuropod_java_tmp",
```
Comment (VivekPanyam) [inline source/neuropod/bindings/java/BUILD:131]: We can entirely remove `neuropod_java_tmp` and just use `neuropod_java` in the genrule above.

(Since we're removing the `tar_backends` dependency here and the dependency on `backends` in `neuropod_java` below, the targets become identical)

File: source/neuropod/bindings/java/README.md [hunk 9]
```diff
@@ -1,6 +1,30 @@
 # Java API
 
 ***NOTE: This API is currently experimental***
+#  Build jar
+```shell script
+bazel build --verbose_failures --action_env=NEUROPOD_TENSORFLOW_VERSION=1.15.0 --action_env=NEUROPOD_TORCH_VERSION=1.4.0 //neuropod/bindings/java:neuropod_java
+```
+
+#  Build fat jar
+Fat jar is self-contained jar that includes native OS-Arch specific files: JNI library, Neuropod Core and Backend packages.
```
Comment (VivekPanyam) [inline source/neuropod/bindings/java/README.md:10]: Please update this comment

**Review by VivekPanyam — COMMENTED** [review]
Looks good overall with a few notes

> I removed "pack backend tar" into Fat Jar part and left only Core native bins in specific directories (for OSX and Linux). Please review.

It looks like there are a few more things to remove. I commented on them below

**Comment by VivekPanyam** [comment]
@vkuzmin-uber Any updates here? Thanks :)
